client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,210 @@
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, callback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, callback );
+			}
+
+			const reqParams = Object.assign( {}, params );
+			const { path } = reqParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, callback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			/**
+			 * Send response immediately to the client
+			 * sendingt the given data
+			 *
+			 * @param {Object} localRecord - response stored locally
+			 */
+			const localResponse = localRecord => {
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					//callback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+			};
+
+			/**
+			 * Handling Error getting data locally
+			 *
+			 * @param {Error} err - err trying to get the local record
+			 */
+			const recordErrorHandler = err => {
+				// @TODO improve error handling here
+				console.error( err );
+			};
+
+			/**
+			 * Fetch data from WP.com
+			 *
+			 * @return {Promise} promise
+			 */
+			const networkFetch = () => {
+				return new Promise( ( resolve, reject ) => {
+					handler( params, ( err, res ) => {
+						if ( err ) {
+							return reject( err );
+						}
+
+						resolve( res );
+					} );
+				} );
+			}
+
+			/**
+			 * Add/Override the data gotten from the
+			 * WP.com server-side response.
+			 * Run the double callback
+			 *
+			 * @param {Object} serverResponse - server response object
+			 */
+			const cacheResponse = serverResponse => {
+				// remove _headers from server response
+				delete serverResponse._headers;
+
+				let storingData = {
+					__sync: {
+						key,
+						synced: new Date().toString(),
+						syncing: false
+					},
+					body: serverResponse,
+					params: reqParams
+				};
+
+				this.storeRecord( key, storingData, err => {
+					if ( err ) {
+						console.error( err );
+
+						if ( ! responseSent ) {
+							callback( err, serverResponse );
+						}
+					}
+
+					// extra response
+					let response = Object.assign(
+						{},
+						serverResponse,
+						{ __syncResponse: true }
+					);
+
+					debug( 'second callback run: %o', response );
+					callback( err, response );
```

Comment:
Clearly we have to define how to handle Errors that aren't completely related with the request/response cycle.
It dependes where the error occurs. In the most of the cases the errors that are not as server responses should be silence them. We don't wanna to notify to the client about that. Maybe we could add them to log system.